File: 2024_YR4/main.py
"""2024 YR4: dual-OD-fit story (early-arc with IP → full-arc IP=0).

Reproduces the 2024yr4 explore-mode scenario from
https://empyrean-dynamics.com/explore/2024yr4.

Run:
    pip install empyrean
    python 2024_YR4/main.py

What it does:
    1. Pulls every available MPC astrometric observation of 2024 YR4.
    2. Filters to the discovery arc (≤ ~2025-01-05) and runs a
       short-arc OD; second pass uses the full arc.
    3. Propagates both fits to the 2032 Earth encounter and reads
       impact probabilities. Early arc surfaces ~0.14% Earth IP with a
       ~12-lunar-distance miss sigma; the full arc collapses it to
       zero. Sentry's published peak was 3.1%
       on 2025-02-18 — different arc cuts give different IPs, all
       legitimately within published uncertainty bounds.

Authoritative cross-checks (printed inline):
    - Discovery 2024-12-27 by ATLAS at Río Hurtado, Chile
    - Sentry peak IP = 3.1% on 2025-02-18 (55-day arc)
    - 2032-12-22 nominal Earth approach: ~278,000 km   (JPL CAD)
    - 2032-12-22 nominal Moon approach:  ~23,000 km    (JPL CAD)
"""

from __future__ import annotations

# spielberg:snippet:start
import numpy as np

import empyrean
from empyrean import Epochs, Origin, TimeScale, UncertaintyMethod


def main() -> None:
    empyrean.initialize()

    # ── 1. Astrometric observations from MPC ────────────────────────
    obs = empyrean.query_observations(["2024 YR4"])
    print(f"{len(obs)} observations from MPC (full arc)")

    # ── 2. Slice to the discovery arc ───────────────────────────────
    # ADESObservations carries obs_time as ISO-8601 UTC strings; convert
    # to MJD via Epochs to filter on time.
    iso_strings = obs.obs_time.to_pylist()
    obs_mjd = Epochs.from_iso(iso_strings, scale=TimeScale.UTC.value).mjd.to_numpy()
    discovery_arc = obs.apply_mask(obs_mjd < 60680.0)  # ~2025-01-05
    print(f"{len(discovery_arc)} observations in the discovery arc (~10 days)")

    # ── 3. Two OD fits ──────────────────────────────────────────────
    def _summary(label: str, r) -> None:
        s = r.summary
        print(
            f"{label}: chi2/dof = {s.reduced_chi2:.3f}  "
            f'RMS RA·cos(d) {s.rms_ra_arcsec:.3f}" Dec {s.rms_dec_arcsec:.3f}"  '
            f"({s.num_selected}/{s.num_obs} obs)"
        )

    early = empyrean.determine(discovery_arc)
    _summary("Discovery arc", early)

    full = empyrean.determine(obs)
    _summary("Full arc     ", full)

    # ── 4. Impact probabilities at the 2032 encounter for both fits ──
    # The covariance-aware IP comes from compute_impact_probabilities —
    # the fitted orbit carries its 6×6 covariance, which is mapped to
    # the encounter B-plane. (Reading sigma_d / IP off the propagation
    # event stream would not work: possible_impact events are purely
    # geometric — they carry the nominal miss distance only.)
    end_epoch = 63730.0  # ~2032-12-31, MJD TDB

    print("\n2032 encounter — impact probability + miss geometry:")
    for label, fit in [("early arc", early), ("full arc", full)]:
        ips = empyrean.compute_impact_probabilities(
            fit.orbit,
            end_epoch,
            [UncertaintyMethod.SECOND_ORDER],
            body_filter=[Origin.EARTH, Origin.MOON],
        )
        body = ips.body.to_pylist()
        miss_km = ips.miss_distance_km.to_numpy(zero_copy_only=False)
        sigma_km = ips.sigma_distance_km.to_numpy(zero_copy_only=False)
        ip_linear = ips.ip_linear.to_numpy(zero_copy_only=False)
        for i in range(len(ips)):
            print(
                f"  {label:9s}  {body[i]:6s}  "
                f"IP_linear = {ip_linear[i]:>7.3%}  "
                f"miss = {miss_km[i]:>10.1f} km  "
                f"sigma_d = {sigma_km[i]:>10.1f} km"
            )

    # Full-arc SecondOrder propagation across the 2032 encounter, used
    # for the tagged-covariance readback in section 5.
    epochs = Epochs.from_kwargs(
        mjd=[61000.0 + 10.0 * i for i in range(271)],
        scale=TimeScale.TDB.value,
    )
    full_prop = empyrean.propagate(
        full.orbit,
        epochs,
        uncertainty_method=UncertaintyMethod.SECOND_ORDER,
        tagged_covariance=True,
    )

    # ── 5. Tagged-covariance readback through the 2032 flyby ─────────
    # The bare linear covariance carried on the propagated states is
    # Phi Sigma0 Phi^T — it ignores the curvature of the dynamics across
    # the encounter. Near a planetary close approach that curvature
    # matters, so we read the *resolved-kind* (Park–Scheeres second-order)
    # covariance back from the full-arc SecondOrder propagation and
    # contrast it with the linear one at perigee.
    peri = full_prop.events.periapses
    bodies = peri.body.to_pylist()
    peri_mjd = peri.epoch.to_numpy(zero_copy_only=False)
    ca_mjd = next(peri_mjd[i] for i in range(len(bodies)) if bodies[i] == "Earth")

    # Resolved-kind covariance, one tag per output epoch. The states
    # table is NOT request-ordered (encounter episodes are grouped by
    # origin), so each table is looked up by its own epoch column.
    series = full_prop.tagged_covariance_series(0)
    out_mjd = np.array([tc.epoch_mjd_tdb for tc in series])
    k = int(np.argmin(np.abs(out_mjd - ca_mjd)))
    state_mjd = full_prop.states.coordinates.epoch.to_numpy(zero_copy_only=False)
    k_state = int(np.argmin(np.abs(state_mjd - ca_mjd)))

    def _pos_sigma_km(m) -> float:
        m = np.asarray(m)
        return float(np.sqrt(m[0, 0] + m[1, 1] + m[2, 2])) * 149_597_870.7

    resolved = series[k]
    resolved_sigma = _pos_sigma_km(resolved.matrix)
    # The bare linear covariance lives on the propagated states.
    linear = full_prop.states.coordinates.covariance.to_matrix()
    linear_sigma = _pos_sigma_km(linear[k_state])

    print(f"\nTagged-covariance readback at 2032 Earth perigee (MJD {ca_mjd:.4f} TDB):")
    print(f"  resolved kind        = {resolved.kind}")
    print(
        f"  resolved pos sigma   = {resolved_sigma:>10.1f} km  (second-order ellipsoid)"
    )
    print(f"  linear   pos sigma   = {linear_sigma:>10.1f} km  (bare Phi Sigma0 Phi^T)")

    print("\nReference (JPL CAD, full-arc nominal):")
    print("  Earth   ~278,000 km   IP = 0")
    print("  Moon    ~23,000 km    IP = 0   (inside Moon's Hill sphere)")
    print("Reference (Sentry, peak):")
    print("  Earth   55-day arc, 2025-02-18 published    IP = 3.1%")


# spielberg:snippet:end


if __name__ == "__main__":
    main()

File: 2024_YR4/main.rs
//! 2024 YR4: dual-fit OD story (early-arc IP vs full-arc ruled-out).
//!
//! Rust twin of `2024_YR4/main.py`. Reproduces the 2024yr4 explore-mode scenario
//! from <https://empyrean-dynamics.com/explore/2024yr4>.
//!
//! Run:
//! ```bash
//! cargo run --release --bin yr4
//! ```

// empyrean:snippet:start
use empyrean::{
    Context, Epoch, ODConfig, Origin, PropagationConfig, UncertaintyMethod, query_observations,
};

const KM_PER_AU: f64 = 149_597_870.7;

fn main() -> empyrean::Result<()> {
    let ctx = Context::from_data_dir(None)?;

    // ── 1. Astrometric observations from MPC ────────────────────────
    let all_obs = query_observations(&["2024 YR4"], None)?;
    println!("{} observations from MPC (full arc)", all_obs.len());

    // ── 2. Strict 10-day discovery arc cut via filter_by_epoch ──────
    // YR4 was first listed on Sentry at 1.2% on 2024-12-27 with the
    // early-arc dataset. The published 3.1% Sentry peak corresponds to
    // the 55-day arc through 2025-02-18. The discovery arc here is
    // the conservative ≤10-day cut.
    let cut = Epoch::from_mjd_tdb(60680.0); // ~2024-12-27 + 10 days
    let early_arc = all_obs.filter_by_epoch(None, Some(cut))?;
    println!(
        "{} observations after 10-day discovery-arc cut (< MJD {:.2} TDB)",
        early_arc.len(),
        cut.mjd()
    );

    // ── 3. Dual OD: discovery-arc vs full-arc ───────────────────────
    let cfg = ODConfig::default();

    let print_summary = |label: &str, r: &empyrean::DetermineResult| {
        let s = &r.summary;
        println!(
            "{label}: chi2/dof = {:.3}  RMS RA·cos(d) {:.3}\" Dec {:.3}\"  ({}/{} obs)",
            s.reduced_chi2, s.rms_ra_arcsec, s.rms_dec_arcsec, s.num_selected, s.num_obs
        );
    };

    let early = ctx.determine(&early_arc, None, &cfg)?;
    print_summary("Discovery arc", &early);

    let full = ctx.determine(&all_obs, None, &cfg)?;
    print_summary("Full arc     ", &full);

    // ── 4. Compute impact probabilities to 2032 encounter for both fits ──
    let end_epoch = Epoch::from_mjd_tdb(63730.0); // ~2032-12-31

    // The fitted orbit is already re-feedable — it carries its covariance
    // and the fitted non-grav model. Pass it straight to the IP call.
    let early_orbit = early.orbit.clone().with_orbit_id("2024 YR4 (discovery)");
    let full_orbit = full.orbit.clone().with_orbit_id("2024 YR4 (full)");

    let print_ips = |label: &str, ips: &[empyrean::ImpactProbability]| {
        println!("\n{label} — 2032 encounter:");
        for ip in ips {
            let miss_km = ip.miss_distance_au * KM_PER_AU;
            let sigma_km = ip.sigma_distance_au * KM_PER_AU;
            println!(
                "  {:?}  IP_linear = {:>7.3}%  miss = {:>10.1} km  sigma_d = {:>10.1} km",
                ip.body,
                ip.ip_linear * 100.0,
                miss_km,
                sigma_km,
            );
        }
    };

    let early_ips = ctx.compute_impact_probabilities(
        &[early_orbit],
        end_epoch,
        &[UncertaintyMethod::SecondOrder],
        &[Origin::Earth, Origin::Moon],
    )?;
    print_ips("Discovery arc", &early_ips);

    let full_ips = ctx.compute_impact_probabilities(
        &[full_orbit],
        end_epoch,
        &[UncertaintyMethod::SecondOrder],
        &[Origin::Earth, Origin::Moon],
    )?;
    print_ips("Full arc", &full_ips);

    // ── 5. Tagged-covariance readback through the 2032 flyby ─────────
    // The bare linear covariance on a propagated state is Φ Σ₀ Φᵀ — it
    // ignores the curvature of the dynamics across the encounter. Near a
    // planetary close approach that curvature matters, so we propagate
    // the full-arc fit through the flyby under SecondOrder uncertainty
    // and read the *resolved-kind* (Park–Scheeres second-order)
    // covariance back, contrasting it with the linear one at perigee.
    let grid: Vec<Epoch> = (0..271)
        .map(|i| Epoch::from_mjd_tdb(61_000.0 + 10.0 * i as f64))
        .collect();

    let prop_cfg = PropagationConfig {
        uncertainty_method: UncertaintyMethod::SecondOrder,
        ..PropagationConfig::default()
    };

    let full_orbit = full.orbit.clone().with_orbit_id("2024 YR4 (full)");
    let prop = ctx.propagate(std::slice::from_ref(&full_orbit), &grid, &prop_cfg)?;

    // Earth periapsis epoch from the detected events.
    let earth_peri = prop
        .events
        .iter()
        .find(|e| e.event_type == "periapsis" && e.body == Some(Origin::Earth))
        .expect("Earth periapsis detected during the 2032 flyby");
    let ca_mjd = earth_peri.epoch.mjd_tdb()?;

    // Resolved-kind covariance series (one tag per output epoch),
    // aligned epoch-for-epoch with `prop.states`.
    let series = prop.covariance_series_cartesian(0)?;
    let k = series
        .iter()
        .enumerate()
        .min_by(|(_, a), (_, b)| {
            let da = (a.epoch.mjd_tdb().unwrap() - ca_mjd).abs();
            let db = (b.epoch.mjd_tdb().unwrap() - ca_mjd).abs();
            da.partial_cmp(&db).unwrap()
        })
        .map(|(i, _)| i)
        .expect("non-empty covariance series");

    let pos_sigma_km =
        |m: &[[f64; 6]; 6]| -> f64 { ((m[0][0] + m[1][1] + m[2][2]).sqrt()) * KM_PER_AU };

    let resolved = &series[k];
    let resolved_sigma = pos_sigma_km(&resolved.matrix);
    // The bare linear covariance lives on the propagated state itself.
    // States are NOT request-ordered (encounter episodes are grouped by
    // origin), so find the row by its own epoch.
    let k_state = prop
        .states
        .iter()
        .enumerate()
        .min_by(|(_, a), (_, b)| {
            (a.epoch.mjd() - ca_mjd)
                .abs()
                .total_cmp(&(b.epoch.mjd() - ca_mjd).abs())
        })
        .map(|(i, _)| i)
        .expect("non-empty states");
    let linear_sigma = prop.states[k_state]
        .covariance
        .as_ref()
        .map(pos_sigma_km)
        .expect("linear covariance present on the propagated state");

    println!("\nTagged-covariance readback at 2032 Earth perigee (MJD {ca_mjd:.4} TDB):");
    println!("  resolved kind        = {:?}", resolved.kind);
    println!("  resolved pos sigma   = {resolved_sigma:>10.1} km  (second-order ellipsoid)");
    println!("  linear   pos sigma   = {linear_sigma:>10.1} km  (bare Phi Sigma0 Phi^T)");

    println!("\nReference (JPL CAD, full-arc nominal):");
    println!("  Earth   ~278,000 km   IP = 0");
    println!("  Moon    ~23,000 km    IP = 0   (inside Moon's Hill sphere)");
    println!("Reference (Sentry, peak):");
    println!("  Earth   55-day arc, 2025-02-18 published    IP = 3.1%");

    Ok(())
}
// empyrean:snippet:end
